<!doctype html>

<head>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta charset="utf-8">
    <!-- Bootstrap v4beta Imports -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/css/bootstrap.min.css" integrity="sha384-/Y6pD6FV/Vv2HJnA6t+vslU6fwYXjCFtcEpHbNJ0lyAFsXTsjBbfaDjzALeQsN6M" crossorigin="anonymous">

    <style media="screen">
        body {
            padding-top: 70px;
            padding-bottom: 70px;
        }
    </style>

    <!-- Syntax Highlighter. Use both pygments and hl.js. -->
    <link rel="stylesheet" href="../../../static/pygments.css">
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css">
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>

    <!-- Google Analytics -->
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-12498603-2', 'auto');
        ga('send', 'pageview');
    </script>

    <!-- ClustrMaps Tracking -->
    <!-- <script type="text/javascript" id="clstr_globe" src="//cdn.clustrmaps.com/globe.js?d=nhKoDpoTjWz4pC6CwI-fSy4hPoJ1uXwTLCfMCT3OK_8"></script> -->

    <!-- FontAwesome embed -->
    <script src="https://use.fontawesome.com/cb9dbe8e41.js"></script>
</head>

<title>Blog - Eric J. Ma's Personal Site</title>
<body class="body">
    <nav class="navbar navbar-expand-sm navbar-light fixed-top bg-light">
        <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#local-links" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
            <span class="navbar-toggler-icon"></span>
        </button>
        <div class="collapse navbar-collapse" id="local-links">
            <ul class="navbar-nav mr-auto">
                
                <li class="nav-item">
                    <a class="nav-link" href="/"><i class="fa fa-home" aria-hidden="true"></i> Home</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="/resume"><i class="fa fa-file-text" aria-hidden="true"></i> Resume</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="/blog"><i class="fa fa-rss" aria-hidden="true"></i> Blog</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="/open-source"><i class="fa fa-code" aria-hidden="true"></i> Open Source</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="/projects"><i class="fa fa-briefcase" aria-hidden="true"></i> Projects</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="/talks"><i class="fa fa-microphone" aria-hidden="true"></i> Talks</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="/teaching"><i class="fa fa-university" aria-hidden="true"></i> Teaching</a>
                </li>
                
            </ul>
        </div>
    </nav>
    <div class="container">
        
  
    
  <!-- <div class="blog-post"> -->
  <link rel="stylesheet" href="../../../static/pygments.css">
  <div>
  
    <h1><a href="../../../blog/2016/11/6/why-phenotypic-interpretation-matters/">Why Phenotypic Interpretation Matters</a></h1>
  
  <p class="meta">
    written by
    
      <a href="https://twitter.com/ericmjl">Eric J. Ma</a>
    
    on 2016-11-06
  </p>
  <p>At IMED 2016, I heard a lot of talks surrounding surveillance efforts. Most of it has been syndromic surveillance, because that's what's currently collectable. For digital disease surveillance, search queries (a marker of "interest" in a disease) are used as a proxy for probable presence of a disease.</p>
<p>These are good for real-time surveillance, but I think it's not good enough from the standpoint of accuracy and understanding mechanism.</p>
<p>For accuracy, syndromic surveillance and search queries are n<sup>th</sup>-degree manifestations of underlying causes. Multiple different pathogens can present the same symptoms. Upticks in searches for a given symptom or pathogen may have multiple underlying causes, though "hearing of the disease" and "being concerned about it" may be related to certain diagnoses and local spread of information.</p>
<p>Found data can help act as a very early sentinel to a problem, and can guide us by providing leads to follow-up, but for understanding mechanism, neither tell us how a new virus is going to be dangerous.</p>
<p>In other words, while fast and rapid and local, syndromic surveillance and digital disease surveillance remains a reactive endeavour, because we can't mechanistically forecast risk from the data.</p>
<p>On the other hand, genomic data is often collected with the intention of understanding pathogen dynamics, as we have well-developed phylogenetic tools for this. However, we still rely on single point mutations, identified in small-scale, one-off, ad-hoc, [insert-some-dashed-phrase-that-expresses-frustration] studies, and basically treat the limited amount of evidence as gospel. I am quite sure that is not the right way to do phenotypic interpretation of genomic data.</p>
<p>If we're going to make disease surveillance a <strong>proactive</strong> endeavour, I am convinced that the disease surveillance world will need a paradigm shift, and investment in infrastructure. By predictive, this is what I mean:</p>
<ol>
<li>Calculating what has a high probability of jumping into humans.</li>
<li>Phenotypically testing those viruses without using live viruses (concept of pseudotyping).</li>
<li>Using genomic information to predict phenotype.</li>
</ol>
<p>Here's how I see it happening, given everything I've learned at IMED 2016:</p>
<p><strong>(1) Sequence everything, in animals that interface with humans</strong></p>
<p>60-70% of new infections are going to be zoonotic in nature. To know what might be coming up, use metagenomics to sequence animal species in geographic areas that humans, <em>prior</em> to the host switch event. To me, this is a no-brainer. Long-term serial sampling is going to be key.</p>
<p><strong>(2) Leverage pseudotyping technologies and genetic systems, and automation, to create systematic phenotyping platforms</strong></p>
<p>Doing so will allow us to phenotypically test components of a virus for a particular step in its life cycle, such as speed of entry into a cell, or resistance to a drug. With rapid DNA synthesis & assembly technologies, we should be able to experimentally measure a virus' epidemiologically-relevant phenotype within 24 hours of an outbreak. We should also be able to rapidly generate large-scale genotype-phenotype data for machine learning.</p>
<p>What would be relevant phenotypes? I can think of a few, broken down basically to "druggability" and "life cycle steps":</p>
<ol>
<li>For druggable proteins, measure drug resistance. It has high epidemiological relevance, and hence, value.</li>
<li>For viruses targetable by drugs and not alike, measure cellular entry capability.</li>
<li>For identified polymerase complexes, measure replication capacity.</li>
<li>For cellular release machinery, if identified, measure release.</li>
<li>If there's others already existing for which the case can be made for systematic phenotyping, scale the assay up.</li>
</ol>
<p><strong>(3) Develop machine learning tools that allow for learning of structural properties that determine epi-relevant protein phenotype</strong></p>
<p>By doing so, we can explicitly model epistasis in a protein, and therefore account for epistatic interactions explicitly when developing learning algorithms that map genotype to phenotype. With this kind of technology coupled to the ability to sequence a virus within hours of sample collection, we'll be able to cut down the time for phenotypic interpretation from the order of a day (because of experimentation required) to seconds (because it's a computational prediction). By going Bayesian, we can even model uncertainty in final predictions, thereby aiding decision-making under uncertainty.</p>
<hr>
<p>In conjunction with the "sequence everything" approach, systematic phenotyping and deep learning will allow us to move to a "predictive" paradigm because it'll give us the ability to rationally predict the phenotypic risk of a virus before it even jumps into humans. I see this as an infrastructural project: requiring up-front commitment of time and money. Yet, we know that investments in infrastructure always pay off dividends in a myriad of ways, from time saved to lives saved.</p>
<p>Something I learned from IMED 2016, however, is that epidemiology is an old discipline, and with old disciplines comes a certain tunnel vision on how things should be done. It'll be up to a new and creative generation (i.e. us) to convince the funders and old guard that our new way will add value to the discipline.</p>

  </div>

    <p><i>Did you enjoy this blog post? <a href="../../../blog/2016/11/6/why-phenotypic-interpretation-matters/#disqus">Let's discuss more</a>!</i></p>
    <hr class="fancy">
  
    
  <!-- <div class="blog-post"> -->
  <link rel="stylesheet" href="../../../static/pygments.css">
  <div>
  
    <h1><a href="../../../blog/2016/11/5/imed-2016-talks/">IMED 2016 - Talks</a></h1>
  
  <p class="meta">
    written by
    
      <a href="https://twitter.com/ericmjl">Eric J. Ma</a>
    
    on 2016-11-05
  </p>
  <p>Today (Day 1) was the day that talks and posters really started. Hackathon participants missed the first day, in which there was a plenary session of "high-level speakers". A bit of a pity to miss it, but that's okay - the hackathon was more fun.</p>
<p>After delivering the Animal Village hackathon pitch once more in the morning with my teammate Emily Iacobucci, I hung around to learn more about what people in the Emerging Infectious Diseases & Surveillance community are thinking about. I was most interested in what set of tools were being developed, and what problems they solved. Below are my summarized notes.</p>
<p><strong>Talks</strong></p>
<p>For the talks, I focused mostly on the experimental and data collection tools being developed. Also attended one talk on forecasting new emerging diseases - this is like the holy grail of infectious diseases - can we predict where new outbreaks are going to happen? Also in development are genetic tools to manipulate <em>any</em> virus. Key tools I took note of included:</p>
<ul>
<li>Number of different viral hosts is a great predictor of spillover into humans. | <a href="http://www.nature.com/articles/srep14830">paper</a></li>
<li>Genetic systems for <a href="https://en.wikipedia.org/wiki/Pseudotyping">pseudotyping</a>. | <a href="https://www.ncbi.nlm.nih.gov/pubmed/25936665">paper (paywall)</a></li>
<li>Syndromic and/or search surveillance software and participatory community surveillance by patients/lay people (<a href="http://www.skollglobalthreats.org/tag/podd/">PODD</a>/<a href="https://flunearyou.org/">Flu Near You</a>) and physicians (<a href="http://www.uptodate.com/home">UpToDate</a>)</li>
</ul>
<p><strong>Posters</strong></p>
<p>For the posters, I mainly caught three posters that dealt with data and modelling; most of the other posters were about some specific outbreak in some specific animal or some specific region. These are all good and such; it's just that my nature inclines me towards generalizable topics:</p>
<ul>
<li>Validation of NA-Fluor assay for clinical use: most interested because of genotype-phenotype problem I'm tackling.</li>
<li>Data sharing by the <a href="https://www.iddo.org/">International Diseases Data Observatory</a>: an interesting one where I learned that data governance has to be done right to get people to share data.</li>
<li>Modelling of disease spread using airplane travel data, by <a href="http://www.ecohealthalliance.org/">EcoHealth Alliance</a></li>
</ul>
<hr>
<p><strong>Update from Day 2</strong></p>
<p>The talks that I picked to hear today were centred mostly on data and antimicrobial resistance (AMR). Highlights include the following:</p>
<ol>
<li>European CDC has a network of AMR surveillance efforts that are going on. I think a lot of it has to be organically grown in individual nations, and can't be forced upon.</li>
<li>Data sharing ethics: Representatives from the Wellcome Trust, WHO, and European CDC gave an overview of the principles behind best open data sharing practices, issues and challenges currently faced, and ethics.</li>
<li>Principles for data sharing:<ol>
<li>Using data to help communities be prepared for an outbreak.</li>
<li>Speed but not at the cost of issues of "justice" (e.g. not inadvertently marginalizing groups in interest of speed)</li>
<li>Identifying appropriate level of risk that society can bear.</li>
<li>Building relationships and confidence.</li>
<li>Moving focus away from ownership of data to providing a fair and just system for data recognition.</li>
</ol>
</li>
</ol>
<p>Apart from this, I sat in on Eddy Rubin's talk on the global virome (more like "stood in", it was standing room only). I am convinced that taking a "sequencing everything" approach is likely going to be the key for predictive surveillance. The next step is whether we'll be able to interpret risk from the genome or not, but that'll be the focus of my next post.</p>

  </div>

    <p><i>Did you enjoy this blog post? <a href="../../../blog/2016/11/5/imed-2016-talks/#disqus">Let's discuss more</a>!</i></p>
    <hr class="fancy">
  
    
  <!-- <div class="blog-post"> -->
  <link rel="stylesheet" href="../../../static/pygments.css">
  <div>
  
    <h1><a href="../../../blog/2016/11/4/imed-2016-hackathon/">IMED 2016 - Hackathon</a></h1>
  
  <p class="meta">
    written by
    
      <a href="https://twitter.com/ericmjl">Eric J. Ma</a>
    
    on 2016-11-04
  </p>
  <p>Yesterday and today, I was at the <a href="http://hackathon.isid.org/">IMED 2016 hackathon</a>, jointly organized by <a href="http://hackingmedicine.mit.edu/">Hacking Medicine at MIT</a> and the <a href="http://isid.org/">International Society for Infectious Disease</a>.</p>
<p>I arrived late for the hackathon because of a cancelled flight, and as such I had to book another flight at the last minute. Because I arrived late (not sure who to blame still: Turkish Airlines or Webjet), I missed out on the opening portion of the hackathon. I joined a team comprised of:</p>
<ul>
<li>Emily Iacobucci, WU Vet School</li>
<li>Bonnie Qian, ETH Zurich</li>
<li>Andres Gaviria, Univ. South Carolina</li>
<li>Raphael Dachs, Vienna University of Economics and Business</li>
</ul>
<p>Our hackathon pitch was as such: To create a "Reddit" or "Quora" or "Stack Overflow" for animal workers, much akin to <a href="https://www.plantvillage.org/">Plant Village</a>. With such a platform, we can help communities help each other with questions about their animals. This also happens to be a great curated knowledge set that can be of great value to governments and veterinary medicine companies.</p>
<p>This being my first hackathon, I learned a number of cool things. Firstly, there's a lot of interesting solutions that can be brainstormed out of a hackathon. For example, there was one about measuring the environmental impact of humanitarian aid, in the form of a CO<sub>2</sub> calculator; it turned out to be the winning idea. (It also was the idea most closely related to the theme of the hackathon, which was climate change and infectious disease.)</p>
<p>Secondly, I learned more about my own work style. Hackathons can be pressurizing scenarios. This can be a polarizing kind of situation to be in; some are not ready for it, others thrive under pressure. I learned something new about myself here - I don't thrive well under pressurizing situations where the goal isn't clear, and it's even harder to bring out my best when I have doubts about the majority's direction. On the flip side, when there's a clear goal forward, that's when my best qualities show up.</p>
<p>The hackathon setting also got me thinking about the kinds of problems that are solved here, versus the kinds of problems solved over a longer period of time. A hackathon environment is especially attuned to forcing out new ideas that are mish-mashes of stuff that already exist, but contrary to my pre-conceived notions, it might not necessarily be the right scenario for solving problems. I think what a hackathon does (that is more impactful than solving problems) is bring people together that might not have necessarily worked with one another before. Very rapidly, we'll figure out what types we're capable of working together with, and what type's we're not.</p>
<p>In contrast, a PhD-timeframe for solving a problem allows us to go much deeper, at a pace that doesn't force out new solutions, but rather at a pace that allows us to execute a series of "personal sprints" that let us move towards a "truth" about the world. The time scales are different, and hence the nature of the problem that we end up solving is much different.</p>
<p>Anyways, we had a great time hacking together as a team, and I think the hackathon was a great time of bonding as well. Our pitch was one of the pitches that will be shown tomorrow morning, so as I conclude this jet-lagged night, it's time to go to bed so that I can be up in time for tomorrow's conference!</p>

  </div>

    <p><i>Did you enjoy this blog post? <a href="../../../blog/2016/11/4/imed-2016-hackathon/#disqus">Let's discuss more</a>!</i></p>
    <hr class="fancy">
  
    
  <!-- <div class="blog-post"> -->
  <link rel="stylesheet" href="../../../static/pygments.css">
  <div>
  
    <h1><a href="../../../blog/2016/11/1/paywalls/">Paywalls</a></h1>
  
  <p class="meta">
    written by
    
      <a href="https://twitter.com/ericmjl">Eric J. Ma</a>
    
    on 2016-11-01
  </p>
  <p>What a twist! A paper that I contributed to is behind a Wiley paywall.</p>
<p><a href="../../../blog/2016/11/1/paywalls/ecoletters.png"><img src="../../../blog/2016/11/1/paywalls/ecoletters-small.png" alt="Ecology letters paywall."></a></p>
<p>(click on the image to enlarge it)</p>
<p>Not the first time it's happened to me, another paper that I helped out with also ended up behind a paywall - the one we wrote on the effect of mutations on influenza polymerase activity.</p>
<p>All I wanted to do for this paper was reference one of the figures inside it.</p>
<p>For these reasons, I think it's very, very important to upload pre-print versions of final papers to pre-print servers - BioRxiv, ArXiv, PeerJ pre-prints etc. Particularly, authors, just before submitting back to the journals after peer review, put an unformatted copy up on BioRxiv first. The content matters much, much more than the formatting (even if the formatting helps with reading).</p>

  </div>

    <p><i>Did you enjoy this blog post? <a href="../../../blog/2016/11/1/paywalls/#disqus">Let's discuss more</a>!</i></p>
    <hr class="fancy">
  
    
  <!-- <div class="blog-post"> -->
  <link rel="stylesheet" href="../../../static/pygments.css">
  <div>
  
    <h1><a href="../../../blog/2016/10/29/pycon-2017/">PyCon 2017</a></h1>
  
  <p class="meta">
    written by
    
      <a href="https://twitter.com/ericmjl">Eric J. Ma</a>
    
    on 2016-10-29
  </p>
  <p>Registered for <a href="https://us.pycon.org/2017/">PyCon 2017</a>, and tutorial & talk proposals submitted!</p>
<p>PyCon remains my favourite annual conference. What makes a conference great really is the people, and the PyCon community remains a great one to be at. Attendees generally know that being welcoming is valued higher than showing off technical prowess. The talks and tutorials are fully recorded and available online, meaning that everything presented there is also available for the rest of the world, <em>for free</em>.</p>
<p>My <a href="https://www.evernote.com/l/AANFXdw6tbFDUY9IjD25sRRGVuxp8fjWvZU">tutorial proposal</a> was on teaching network analysis. This will be my 3rd year showing Pythonistas how to use NetworkX as a tool for their modelling problems.</p>
<p>My <a href="https://www.evernote.com/l/AAOw0Y3-pzdAdKJUgX3tNMUGj5_H2L6vG-4">talk proposal</a> was on the use of PyMC3 to do Bayesian statistical modelling. My goal is to show how one doesn't really need to know deep math to wield the tools necessary to do Bayesian statistical analysis, even though I will concede that having a strong grasp of the underlying mathematical theory is <em>very</em> helpful.</p>
<p>Hoping to have a blast at next year's PyCon! The timing also happened to be just right, just before MIT's convocation ceremony, and (hopefully) a month after my thesis defence. PyCon 2017, here we come!</p>

  </div>

    <p><i>Did you enjoy this blog post? <a href="../../../blog/2016/10/29/pycon-2017/#disqus">Let's discuss more</a>!</i></p>
    <hr class="fancy">
  

  
  <div class="pagination">
    
      <a href="../../../blog/page/10/">&laquo; Previous</a>
    
    | 11 |
    
      <a href="../../../blog/page/12/">Next &raquo;</a>
    
  </div>


    </div>

    <nav class="navbar navbar-expand-sm navbar-light fixed-bottom bg-light">
        <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#external-links" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
            <span class="navbar-toggler-icon"></span>
        </button>
        <div class="collapse navbar-collapse" id="external-links">
            <ul class="navbar-nav mr-auto">
                
                <li class="nav-item">
                    <a class="nav-link" href="https://www.linkedin.com/in/ericmjl"><i class="fa fa-linkedin" aria-hidden="true"></i> LinkedIn</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="http://twitter.com/ericmjl"><i class="fa fa-twitter" aria-hidden="true"></i> Twitter</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="http://github.com/ericmjl"><i class="fa fa-github" aria-hidden="true"></i> GitHub</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="http://stackoverflow.com/users/1274908/ericmjl"><i class="fa fa-stack-overflow" aria-hidden="true"></i> Stack Overflow</a>
                </li>
                
                <li class="nav-item">
                    <a class="nav-link" href="http://shortwhale.com/ericmjl"><i class="fa fa-envelope-o" aria-hidden="true"></i> Contact Me</a>
                </li>
                
            </ul>
        </div>
    </nav>
    <!-- Boostrap JS imports -->
    <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.11.0/umd/popper.min.js" integrity="sha384-b/U6ypiBEHpOf/4+1nzFpr53nxSS+GLCkfwBdFNTxtclqqenISfwAzpKaMNFNmj4" crossorigin="anonymous"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/js/bootstrap.min.js" integrity="sha384-h0AbiXch4ZDo7tp9hKZ4TsHbi047NrKGLO3SEJAg45jXxnGIfYzk4Si90RDIqNm1" crossorigin="anonymous"></script>

</body>
